integration/instance.go

```diff
@@ -0,0 +1,392 @@
+	}
+	obj.tmpPrefixDirectory = tmpPrefix
+
+	testRootDirectory := path.Join(obj.dir, RootDirectory)
+	if err := os.MkdirAll(testRootDirectory, dirMode); err != nil {
+		return errwrap.Wrapf(err, "can't create instance root directory")
+	}
+	obj.testRootDirectory = testRootDirectory
+
+	obj.convergedStatusFile = path.Join(obj.dir, ConvergedStatusFile)
+
+	return nil
+}
+
+// Close cleans up after we're done with this instance.
+func (obj *Instance) Close() error {
+	if !obj.Preserve {
+		if obj.dir == "" || obj.dir == "/" {
+			panic("obj.dir is set to a dangerous path")
+		}
+		if err := os.RemoveAll(obj.dir); err != nil { // dangerous ;)
+			return errwrap.Wrapf(err, "can't remove instance dir")
+		}
+	}
+	obj.Kill() // safety
+	return nil
+}
+
+// Run launches the instance. It returns an error if it was unable to launch.
+func (obj *Instance) Run(seeds []*Instance) error {
+	if obj.cmd != nil {
+		return fmt.Errorf("an instance is already running")
+	}
+
+	if len(seeds) == 0 {
+		// set so that Deploy can know where to connect
+		// also set so that we can allow others to find us and connect
+		obj.clientURL = "http://127.0.0.1:2379"
+		obj.serverURL = "http://127.0.0.1:2380"
+	} else {
+		// pick next available pair of ports
+		var maxClientPort, maxServerPort int
+		for _, instance := range seeds {
+			clientPort, err := ParsePort(instance.clientURL)
+			if err != nil {
+				return errwrap.Wrapf(err, "could not parse client URL from `%s`", instance.Hostname)
+			}
+			if clientPort > maxClientPort {
+				maxClientPort = clientPort
+			}
+
+			serverPort, err := ParsePort(instance.serverURL)
+			if err != nil {
+				return errwrap.Wrapf(err, "could not parse server URL from `%s`", instance.Hostname)
+			}
+			if serverPort > maxServerPort {
+				maxServerPort = serverPort
+			}
+		}
+		if maxClientPort+2 == maxServerPort || maxClientPort == maxServerPort+2 {
+			return fmt.Errorf("port conflict found")
+		}
+
+		obj.clientURL = fmt.Sprintf("http://127.0.0.1:%d", maxClientPort+2) // odd
+		obj.serverURL = fmt.Sprintf("http://127.0.0.1:%d", maxServerPort+2) // even
+	}
+
+	cmdName, err := BinaryPath()
+	if err != nil {
+		return err
+	}
+	cmdArgs := []string{
+		"run", // mode
+		fmt.Sprintf("--hostname=%s", obj.Hostname),
+		fmt.Sprintf("--client-urls=%s", obj.clientURL),
+		fmt.Sprintf("--server-urls=%s", obj.serverURL),
+		fmt.Sprintf("--prefix=%s", obj.tmpPrefixDirectory),
+		fmt.Sprintf("--converged-timeout=%d", convergedTimeout),
+		"--converged-timeout-no-exit",
+		fmt.Sprintf("--converged-status-file=%s", obj.convergedStatusFile),
+	}
+	if len(seeds) > 0 {
+		urls := []string{}
+		for _, instance := range seeds {
+			if instance.cmd == nil {
+				return fmt.Errorf("instance `%s` has not started yet", instance.Hostname)
+			}
+			urls = append(urls, instance.clientURL)
+		}
+		s := fmt.Sprintf("--seeds=%s", urls[0])
+		// TODO: we could just add all the seeds instead...
+		//s := fmt.Sprintf("--seeds=%s", strings.Join(urls, ","))
```

Comment:
Since we will be testing the behaviour of how instances in cluster behave given different configurations of seeds either way can be expected to work the same. Is there any (speed) benefit to giving all seeds?

Reply:
I don't think there's a performance difference, in general it's preferable to connect with more seeds in case one happens to go down before you get the full list. Once you connect to anyone, you get the full list. I left it this way since when I refactor etcd/etcd.go I want to make sure there's no bug here, and I wanted to remind myself to at least see if there's a need to test both or not.